test/query.test.js

```diff
@@ -4313,4 +4313,30 @@ describe('Query', function() {
     assert.strictEqual(found.length, 1);
     assert.strictEqual(found[0].title, 'burrito bowl');
   });
+
+  it('update operation should remove fields set to undefined (gh-12794)', async() => {
+    const m = new mongoose.Mongoose();
+
+    await m.connect(start.uri);
+
+    const schema = new mongoose.Schema({ title: String });
+
+    const Test = m.model('test', schema);
+
+    const doc = await Test.create({
+      title: 'test'
+    });
+
+    assert.strictEqual(doc.title, 'test');
+
+    const updatedDoc = await Test.findOneAndUpdate(
+      {
+        _id: doc._id
+      },
+      { title: undefined },
+      { returnOriginal: false }
+    );
+
+    assert.ok(updatedDoc.title === undefined);
```

Comment:
`assert.ok(('title' in updatedDoc) === false);` is failing the test, but it fails also using `save()`:
```
article.title = undefined;
await article.save();

console.log(article.title); // undefined
console.log('title' in article); // true
```
If you think that `title` should not be returned at all we should probably fix both operations for consistency.

Reply:
@lpizzinidev is right, because of how Mongoose works `'title' in article` is always true, even if `title` is undefined. That's because Mongoose always has a getter/setter for every field in the schema.

Better alternative would be to make the `findOneAndUpdate()` use `lean()`, then `assert.ok(!('title' in article))` will work.